staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/cel_validation.go

```diff
@@ -0,0 +1,250 @@
+	Schema *structuralschema.Structural
+	// DeclType is a CEL declaration representation of Schema of this CelSchemaContext node. It is null if the current schema
+	// has no x-kubernetes-validations rules or if the schema is not a valid structural schema.
+	DeclType *model.DeclType
+
+	// Converter converts a JSON schema into a structuralschema.Structural and model.DeclType.
+	Converter func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*structuralschema.Structural, *model.DeclType, error)
+
+	// MaxCardinality represents a limit to the number of data elements that can exist for the current
+	// schema based on MaxProperties or MaxItems limits present on parent schemas, If all parent
+	// map and array schemas have MaxProperties or MaxItems limits declared MaxCardinality is
+	// an int pointer representing the product of these limits.  If least one parent map or list schema
+	// does not have a MaxProperties or MaxItems limits set, the MaxCardinality is nil, indicating
+	// that the parent schemas offer no bound to the number of times a data element for the current
+	// schema can exist.
+	MaxCardinality *uint64
+
+	// TotalCost accumulates the x-kubernetes-validators estimated rule cost total for an entire custom resource
+	// definition. A single TotalCost is allocated for each validation call and passed through the stack as the
+	// custom resource definition's OpenAPIv3 schema is recursively validated.
+	TotalCost *TotalCost
+}
+
+// Converter converts from JSON schema to a structural schema and a CEL declType, or returns an error if the conversion
+// fails.
+type Converter func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*structuralschema.Structural, *model.DeclType, error)
+
+// DefaultConverter converts using standard conversion utilities.
+var DefaultConverter = func(schema *apiextensions.JSONSchemaProps, isRoot bool) (*structuralschema.Structural, *model.DeclType, error) {
+	structural, err := structuralschema.NewStructural(schema)
+	if err != nil {
+		return nil, nil, err
+	}
+	return structural, model.SchemaDeclType(structural, isRoot), nil
+}
+
+// RootCelContext constructs CelSchemaContext for the given root schema.
+func RootCelContext(schema *apiextensions.JSONSchemaProps, converter Converter) CelSchemaContext {
+	rootCardinality := uint64(1)
+	r := CelSchemaContext{
+		MaxCardinality: &rootCardinality,
+		TotalCost:      &TotalCost{},
+		Converter:      converter,
+	}
+	r.initSchemaFields(schema, true)
+	return r
+}
+
+// ChildFieldContext constructs and returns a CelSchemaContext for fieldName.
+func (c *CelSchemaContext) ChildFieldContext(schema, propSchema *apiextensions.JSONSchemaProps, fieldName string) CelSchemaContext {
+	r := c.childContext(schema)
+	if c.Schema != nil && c.Schema.Properties != nil {
+		if propSchema, ok := c.Schema.Properties[fieldName]; ok {
+			r.Schema = &propSchema
+		}
+		if declField, ok := c.DeclType.Fields[fieldName]; ok {
+			r.DeclType = declField.Type
+		}
```

Comment:
am I reading correctly that we always use the `schema` param to build childContext, but in this code path, we ignore the `propSchema` parameter? is there a way to have an initialization step and a consumption step, so we don't leave ourselves code paths where we ignore inputs based on cached data?

Reply:
similar comment about sometimes ignoring params in ChildValuesContext and ChildItemsContext functions below

Reply:
I've split initialization and consumption. This ended up being a substantial change. I think it's a net win since initialization becomes really simple and I was able to a lot of logic to a couple routines which I find easier to reason about. Let me know what you think.

Reply:
I really like the new structure